source/common/upstream/upstream_impl.cc

```diff
@@ -62,6 +62,13 @@ namespace Envoy {
 namespace Upstream {
 namespace {
 
+std::string addressToString(Network::Address::InstanceConstSharedPtr address) {
+  if (!address) {
+    return "";
+  }
+  return address->asString();
+}
```

Comment:
I'm concerned that there is a lot of code that probably assumes that hosts always have addresses. Did you consider making some kind of synthetic address as a sentinel for the proxied case? That might also make a lot of the special cases in this PR simpler?

Reply:
We already have to handle nullptr addresses for UDS, which is why the fixes were in the DynamicCluster code and not generic upstream code.  I'm inclined to think we're better off handling null addresses than having something synthetic, but I added a runtime guard in case the current testing wasn't sufficient (rather than relying on Lyft's already off-by-default rollout of the proxy code)

Reply:
Oh also this is largely a latency/perf fix.  It's generally a latency win doing a DNS lookup for an endpoint we're not connecting to (proxy code connects to proxy address) but it's also possible that folks setting up proxying on a corp network would block DNS so this would make requests work where the might not otherwise.